### PR TITLE
[MoE][Pytorch]Fix size mismatch error in fp8 transpose.

### PR DESCRIPTION
**Pull request by Victarry**
# Description

Replace the return value from input with output in `fp8_cast`.

When training MoE models and disable gradient accumulation fusion, if an linear expert receive zero number of tokens, this bug will cause tensor shape mismatch between wgrad and weight.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
Seems reasonable to me.
